### ISSUE TITLE
Starting a new training run with ladder code

### ISSUE THREAD

**Issue by featurecat** [leela-zero/leela-zero#2259]
Ever since LeelaZero was released, people have been wanting to use it to analyze their games, and it's become incredibly easy and popular to review with LeelaZero. Now, LeelaZero has become insanely strong -- so strong that gaining even a 60-90% winrate over the current network would not significantly improve the usefulness of the program, especially from a reviewing games perspective. **Continuing this run is almost pointless now.**

Now would be a good time to restart a new training run with different parameters. Consistently throughout training, LeelaZero has been weak to ladders. LeelaZero still misjudges ladders all the time. From a review perspective, this is not acceptable. **Go players need to doubt Leela's analysis for every move, because its analysis surrounding ladders will often be wrong.**

You might argue that we should keep increasing LeelaZero's strength past the current network 206, because it will bring us closer to the 'truth' of Go. However, because LeelaZero needs at a minimum 100 playouts to evaluate ladders, and sometimes even over 50,000 to spot a ladder, LeelaZero will never have a chance. **The current run of LeelaZero can never truly understand Go.** Its joseki and fuseki have been incorrectly built without knowledge of one of the most fundamental concepts of Go.

Another benefit of ladders would be that we could finally have networks that give an even game to players of every level. **Users cannot have fair, even games against LeelaZero due to the ladder issue.** Because of the easily exploitable ladder issue, human players from 15 kyu to even dan level can easily win games against LeelaZero, even when they are not intentionally exploiting the ladder weakness. By fixing the ladder issue, LeelaZero could revolutionize how players learn to play Go, at every rank.

For the Go community a new training run with ladder code would be the most useful addition to LeelaZero. Thanks for reading.

**Comment by iopq** [comment]
I'm more interested in how to make it learn ladders by itself. What is with the current approach that makes it only look at the more local situation, and often miss the global picture?

Maybe a stronger bot that is better at global reading doesn't have a preference for 3-3 invasions, for example

**Comment by alreadydone** [comment]
I'm all for introducing ladder features, but I don't agree with starting a new run. Due to our limited resources, we'd like to reuse/transfer current achievements as much as possible. Feel free to suggest Minigo start a new run with ladder features!

I think the best way to go is to add ladder features to the current 40b net using @Ttl's net2net script in his branch https://github.com/Ttl/leela-zero/tree/liberty_planes_se2. The branch also adds liberty and legality features (which are cheap, so why not add them all? I think all these features will help increase strength, reduce blind spots, and helped Master, a 20b net, achieve high strength.) and adopts squeeze-excitation architecture (which may or may not be desirable). Though the branch is dated it shouldn't be hard to extract the useful part. Then we just need to convert a window of training data to start training the net2net'd network, and release a new engine to start self-play with it.

**Comment by MarcelGruenauer** [comment]
I recently had the opportunity to speak to someone who worked on Golaxy. He told me - and I'm a bit fuzzy on the details - that Golaxy gets ladders right even with one playout. He said it doesn't use just the normal policy and value networks but actually four networks; one is for the current score, and the other... I forgot.

The mobile app doesn't run the network on the phone but always uses a server; it has levels from 7k to 9d and all of those use just one playout. There are stronger levels that use more playouts but to be able to play those you first have to defeat the normal 9-dan version under match conditions.

All this is free, and in a few months, they want to release an English version of their app.

Not a very useful comment, I know, but maybe someone knows more about how additional networks, especially for keeping the score, might work or help.

**Comment by Glrr** [comment]
I disagree for several reasons. 
 
 Besides ladders, LZ has still a lot of weak points. With 100 playouts (or even 1k playouts), which should be enough to ensure a pro level, I saw her miss moves that would be obvious for a 1d. Hence training the current network is still useful.

 Even if we add a ladder feature. There is no need to start from scratch. We can start from the recent training games.

  Finally, I think that the feature that is the most needed for game analysis is a score estimation. Instead of a new run of LZ with ladder feature, I would prefer a run of Saigo on 19x19.

**Comment by Vandertic** [comment]
@Glrr SAI 19x19 will be starting its first run in a couple of months.

**Comment by herazul** [comment]
@featurecat I use LeelaZero with my go teacher who love to study new move from AlphaGo games or pro meta games (positions on the board that are very interesting but where there is still no clear best move, or where AlphaGo played something and we still don't know why it played that)
And we definitly observe that the newer the net, the best solution we have from it. There were move that AlphaGo played in certain position that LZ didn't even considered a few net ago, and that she can find and study now.

This is the example from last week study : 
![image](https://user-images.githubusercontent.com/1439756/53484358-f3fbc100-3a83-11e9-9cf8-162f3ababeb9.png)
This is a very interesting board position, and the move in blue is very hard to analyze. LZ now manage to found and read it.

There is a lot of really hard board positions that she read better and faster as nets promotes.


So this is wrong : 
>>gaining even a 60-90% winrate over the current network would not significantly improve the usefulness of the program, especially from a reviewing games perspective.

This affirmation is a gross exageration : 
>>Go players need to doubt Leela's analysis for every move, because its analysis surrounding ladders will often be wrong.
Could you source it ? Can you share board positions where LeelaZero 206 is wrong when you analyze it ?

This is just false : 
>>However, because LeelaZero needs at a minimum 100 playouts to evaluate ladders, and sometimes even over 50,000 to spot a ladder, LeelaZero will never have a chance. **The current run of LeelaZero can never truly understand Go.**

It's almost certain now that AlphaGo take more time than LZ to read ladders. No zero bots "understand" ladder from one playout (without additional tricks like added plane), but they are still the best bots to find the best GO moves. You just have to let it run for more than 100 or 1000 playout, obviously.

Doing a training run with some kind of tricks to detect ladder may not result in a better net, we have no proof of that, and it may definitly not be the case.

Ladders aren't a problem at all if you analyze some moves with a decent computer. (keep in mind that even if LZ knew ladders, from 1 playout you would maybe have a better analysis on some move at low playout, but it would be a low quality analysis either way.
The best thing you can do if you want good analysis is to invest in a decent computer that can give you thousands and thousands of playouts.
With current AI tech, no bot in the world will give you a good analysis of a complex board position with 100 or 1000 playouts, not even AlphaGo.

**Comment by Vandertic** [comment]
Thank you @herazul! Very interesting insight. This kind of analysis and feedback is really useful.

**Comment by croosn** [comment]
@featurecat Hey, so I'd like to express my gratitude for making Lizzie, it's proven to be very successful so far and is used by many pros around the world. And I am, to my disappointment, the first one here to back you up on this. I'd definitely love to see this feature. I remember you saying, when you first started Lizzie project, that your goal was to have an equal opponent for every rank. Which is a wonderful idea if you ask me. The only thing now stopping you from that is custom made ladder detection. I remember seeing some guys, amongst other bearcat (one of the authors of Minigo, I remember seeing him here on github too), today on discord talking about ladder detection feature. Are you working together on this? Anyways, I'm looking forward to seeing this so much. As a high dan player, it's very annoying to rely on any network so far, just to fail to simple ladders... this way any game analysis can be very annoying, and I've heard this from many strong players. Though, I must admit, networks starting from #205 are quite good with ladders. Please keep this idea alive, I bet there are lots of guys like me who would love to see it live. Are you planning on implementing it in Lizzie?

@alreadydone 's idea of using @Ttl 's solution is also nice, though it might miss many ladders... at least that's what I've found out.

@herazul here's two games. Can any bot see that the last move is a ladder breaker? Zen 7 can. And remember that many of us have crappy laptop and limited time and resources.
Game 01 https://ufile.io/hn7b3
Game 02 https://ufile.io/onwwr

![game ladders 01](https://user-images.githubusercontent.com/38720539/53508013-44414600-3ab9-11e9-8811-0ec1d7a67517.jpg)
![game ladders 02](https://user-images.githubusercontent.com/38720539/53508025-47d4cd00-3ab9-11e9-903b-0a1402bdf697.jpg)

**Comment by alreadydone** [comment]
re: score estimation
@pangafu has trained several "OZ" networks which takes into account the score of close games, so the network can predict the winrate under different komi; the winrate estimation is becoming accurate ina reasonable komi range. It's then easy to perform a binary search to estimate the score as the komi that leads to a 50-50 winrate. I did this in muy dynamic komi branches and other people has also implemented it.

If you want the likelihood of each intersection on the board to be occupied by black or white, then neither this nor the SAI approach is able to do it.

**Comment by herazul** [comment]
@croosn I'm not sure what you mean by "Can any bot see that the last move is preparing a ladder? Zen 7 can" so i'm not sure what to test in lizzie.
Could you elaborate a little what your expected result is, what zen7 analyze, and why other bot results doesn't satisfy you ?
I want to understand what you meant to test in lizzie but i don't quite get it.

**Comment by croosn** [comment]
@herazul 
Game 01: if W plays any "normal" or suggested move, B will go out with Q13, followed by sharp win percentage drop.
Game 02: B can play H12 and catch W group in ladders. Even if you force it, LZ doesn't see the continuation.

**Comment by alreadydone** [comment]
> @alreadydone 's idea of using @Ttl 's solution is also nice, though it might miss many ladders... at least that's what I've found out.

Can you clarify? Did you try that branch and find that it failed to recognize some ladders (any one-liberty-two-liberties-alternating scenario)? To be honest I haven't carefully looked at the code to see how things were implemented.

**Comment by croosn** [comment]
Yes, I got the binary from someone couple of months ago, and tested various positions. Shicho still happens.

**Comment by alreadydone** [comment]
@Ttl's branch is meant to be used with a network with added input feature planes which can't be loaded by current official leelaz. Chances are that what you tried was actually the branch https://github.com/leela-zero/leela-zero/pull/1941.

Although Ttl trained such a network from supervised learning, the training data have not covered enough ladder situations, and no reinforcement learning from self-play has been done with such a network. So if you indeed tried Ttl's branch with his network, it's not surprising that it doesn't yet handle all ladders well. However, since it's much easier for such a network to learn ladders, when it's properly trained it would be able to detect ladders in 0 playouts and devote other playouts to something more meaningful.

**Comment by croosn** [comment]
Yes, I used both the branch and two networks by Ttl 10b and 20b. But if you're confident about quick detection, then just great! I'm not an expert in this field, just a messenger of tested stuff.

**Comment by wonderingabout** [comment]
my thoughts : 

- restart from scratch, good idea, a 20b size from the start seems a good idea, and indeed lz is already more than strong enough, BUT dont do a new run until fixed the known problematic blindspots leela zero currently has and other AI dont have as much as leela-zero (until then, a new run will be as bad as this one)
as people said, ressource are limited, so before experimenting, fix the problems first, or it will all go to waste
- for ladders : play a ladder is not forbidden by the rules, so if we want to stick with zero training, ladder detection should not be introduced (or call it leela-one, not leela-zero)
ladder detection however can be useful especially for very low playouts for mobile phones, laptops, weak gpu, etc...
but not in training

**Comment by Hersmunch** [comment]
I agree that it would be nice for leelaz to have support for networks with ladder features as it is an obviously a feature that is of interest. Most of all I like the idea that it keeps more people in the community focused on one code base that is useful for everyone and so everyone benefits from the most recent commits.
However, I think that if we do do some training with a new network architecture then it would be best to only use some of the community's resources rather than all. Well, at least until it is proven to some degree that can be agreed upon.

**Comment by l1t1** [comment]
insteresting

**Comment by tapsika** [comment]
A few points:

1. The current "zero" bots are not really zero. They omit features like liberties, legality and ladder, but at the same time have the "age" feature for stones, encoded in a way that maximises the NNs involvement with it. The idea that a bot should focus attention near fresh stones so the output for the same board position can be different depending on this move order data - this is a human idea. (Even if you think this is necessary for detecting long cycles, in practice it is mostly used for other things.) On the other hand "move legality" would not violate zeroness.

2. For ladder, I think it would be best to add "color and distance of closest stone" in all directions. This would still make it half-zero, only aiding the net with visibility problems.

BTW: it may be possible to add ladder to current LZ nets as well, without changing the codebase. This could work by initializing a net in a way that some useful features (like the above) are precomputed, then start training from that point.

3. Given the inertia of new runs, I agree that would only make sense when all problem points are sorted out, and the community is ready for a run for max strength. This should also include things like score prediction/maximisation, better search code without blind spots (it's wrong to go above 10-20k sims without a single visit for all legal moves), excluding illegal moves from the loss function, etc.

**Comment by Strappa71** [comment]
Would it be sensible to "aid" the bot with things that can be proved as a fact as opposed to human "opinion"?

I am thinking about stuff like symmetry or the fact that a ladder works or not. In that sense the network can focus on opinion of which move is best.

It would of course not be true "zero" in the Alpha Zero way but still zero human strategy input.

**Comment by iopq** [comment]
The bigger question is why it doesn't learn these things from self play...

Symmetry is one thing, but a ladder is a certain pattern

**Comment by featurecat** [comment]
@croosn It's encouraging to hear that I'm not alone. If GCP is not on board with this idea, I will develop my own patch that I will release with Lizzie 0.7. It won't be a perfect fix since Leela Zero might have settled on better joseki if it had known about ladders (it would be a great experiment to see how a Zero bot learns fuseki differently when it has ladders), but it will provide quality analysis.

Those ladder examples are good. I will make sure to test them after my ladder fix.

**Comment by featurecat** [comment]
@iopq Even if a network learned to read out ladders like our current networks (mostly) do, there's another problem. At a minimum it takes 50-100 playouts to read out a ladder. And because the network doesn't have memory of whether or not the ladder worked, it has to read out these 50-100 playouts many times if it is examining multiple different variations. Especially with just 1600 visits to use in self play, when there is a simple ladder on the board, I wonder what percentage of playouts must be devoted to reading the ladder instead of reading important, high level moves? I think it would be tremendous for self-play and in general if the network knew the result of the ladder instantly.

Symmetry is one thing, ladder is a pattern; but it's an extremely useful pattern to include if you want a program that humans can use to review their games at a high level.

**Comment by Hersmunch** [comment]
One bodge/workaround that might work is by using the commands added in #1949 in a gui/analysis tool with a ladder detector added to avoid starting ladders that fail. This probably would only be useful at the root node though as it is not possible to avoid a particular sequence of moves (e.g. only avoid one move if another has been played).

**Comment by alreadydone** [comment]
https://arxiv.org/abs/1902.10565
> With our techniques, in only a week with several dozen GPUs it achieves a likely strong pro or perhaps just-super-human level of strength. Compared to Leela Zero, we estimate a roughly 5x reduction in self-play computation required to achieve that level of strength

Whoa it's a big claim in this paper; maybe a new run wouldn't be that expensive after all.
https://github.com/lightvector/KataGo#katago
https://blog.janestreet.com/accelerating-self-play-learning-in-go/

**Comment by l1t1** [comment]
nice paper

**Comment by iopq** [comment]
@featurecat there must be a neural net that can just detect ladders by looking at them. You don't need to lay them out, just look diagonally across the board. Why didn't lz develop this kind of intuition?

**Comment by featurecat** [comment]
@iopq Ladders inevitably require some reading. Coosm's second ladder example is a good one. So it makes sense that Leela Zero reads out the ladders via playouts. It has a tendency to give up on them early which is the confusing part for me. Maybe it has a hard time balancing the complications of a ladder's policy and value network, with the rest of the board's policy and value network.

**Comment by featurecat** [comment]
@hersmunch I didn't realize that PR was eventually merged! That's great, Lizzie will definitely use that for its version 0.7 (allowing users to specify locations to investigate/not investigate, deny early 3-3, etc). It might also be a quick fix for ladders, although to really fix this it has to be done at the tree search level, by modifying LeelaZero.

**Comment by featurecat** [comment]
Here is an example of how the latest Leela Zero network completely fails to evaluate ladders. 

After many playouts, it still hasn't found Black's correct move. 
![screenshot_20190301_134618_com google chromeremotedesktop](https://user-images.githubusercontent.com/10515890/53659108-aec4c400-3c28-11e9-9ec4-3e4c54304728.jpg)

Once black plays it, it takes about 2000-2500 playouts to start to notice that white can't run out of the ladder. Then it found a point in the upper left that counters this variation at >56% winrate, so Leela Zero would probably stop analyzing the ladder variation at this point. 
![screenshot_20190301_134726_com google chromeremotedesktop](https://user-images.githubusercontent.com/10515890/53659221-fba89a80-3c28-11e9-8261-833b114749a6.jpg)

However if you look at what LZ is actually reading, you'll see that it has miscalculated the ladder a second time. 
![screenshot_20190301_134732_com google chromeremotedesktop](https://user-images.githubusercontent.com/10515890/53659331-4aeecb00-3c29-11e9-8a9f-afffbb0c5e29.jpg)

In fact, there are dozens of variations in this position where LZ needs several thousand playouts to realize the white stones are in a ladder. That means LZ can incorrectly evaluate variations with ladders, requiring tens of thousands of playouts before it begins to correctly analyze them. Often LZ misses ladder variations entirely because the MCTS thinks these variations are not worth investigating further, due to these miscalculations requiring tens of thousands of playouts to correct.

With some ladder code, these variations could be read accurately with orders of magnitude fewer playouts.

**Comment by nerai** [comment]
@featurecat Can you confirm that in your last image, when black plays at 11 instead of at 6, his winrate actually increases?

I am asking because black might believe the ladder works but is just not the best move to play. For instance because of a ladder breaker (unlikely in this example) or due to the atari at 8 which could split off his 2 stones on the right. I don't really think it's the case, just wanted to be sure.

**Comment by iopq** [comment]
@featurecat in the general case you need to read out ladders, but in most cases just drawing two diagonal lines solves your ladder question

but LeelaZ can't even imagine a ladder to the end of an empty board for some reason

**Comment by featurecat** [comment]
@nerI confirmed.
![screenshot_20190301_162121_com google chromeremotedesktop](https://user-images.githubusercontent.com/10515890/53667201-4aacfa80-3c3e-11e9-9a2e-933cfdf6234a.jpg)

After playing the move, and realizing the ladder doesn't work, LZ has the same problem as before. It uses many playouts on another move and forgets that the ladder works. 

![screenshot_20190301_162130_com google chromeremotedesktop](https://user-images.githubusercontent.com/10515890/53667243-7203c780-3c3e-11e9-8de3-63d6bbe07c10.jpg) 

![screenshot_20190301_162203_com google chromeremotedesktop](https://user-images.githubusercontent.com/10515890/53667260-81831080-3c3e-11e9-9bd4-d1513a8b7756.jpg)

As you can see this is quite an extensive problem that exists in most variations.

**Comment by roy7** [comment]
Although there's some papers about trying to re-use information like this, I don't think any current AI projects do so. It has to be re-searched over and over each move and each variation. :(

**Comment by featurecat** [comment]
Correct but with ladder code this would be a few hundred playouts to read instead of many thousands. No need for information re-use.

**Comment by Splee99** [comment]
I guess "Alphago zero learned ladder only in later stage" is a weak statement. Perhaps Alphago zero still has ladder issue, but may be covered by millions play out per move. "Special crafted feature" for ladder should be a necessary amendment for the zero approach.

**Comment by tapsika** [comment]
Actually they claimed that ladders were only "understood" in later stage. This wording likely indicates that their net did develop the necessary long distance connections to guess most ladders on sight (which is important for deeper nodes during search).

That guess would be easy with the extra vision provided by nearest stone feature.

**Comment by bernds** [comment]
Just thinking out loud... has anyone thought of adding additional input planes containing reading results for questions like "can this group with two liberties be captured in a forced sequence?" So, board positions after the played out ladder, when liberties reach zero or more than two, and maybe an indication of which move started it. Ladders are really only a special case, something like this is really the same kind of issue:
https://lifein19x19.com/viewtopic.php?f=18&t=16437
Of course the immediate downside is that it might require a lot of planes.

**Comment by iopq** [comment]
The reason it requires 200k+ playouts is the move that Leelaz likes usually stays the most searched despite better moves already being explored.

I don't know why this behavior happens. It's always obvious after a few thousand playouts the move with the higher win rate is better, but it keeps going with its original guess anyway

**Comment by croosn** [comment]
In rare occasions, it is good to play out ladders that are not working. White was FineArt giving pro player two stones.
Example 03: https://ufile.io/gel90
![ladders 03](https://user-images.githubusercontent.com/38720539/53703243-feb1af80-3e0f-11e9-9fd1-19124b2e1940.jpg)

**Comment by featurecat** [comment]
That's a good point. In my fix I wasn't planning to ban broken ladders; just to better inform the tree search that they exist.

**Comment by pliuphys** [comment]
@iopq you can change the parameter c_puct to achieve that. But the value currently used is thought to be optimal. I tried changing it, only to find the bot got weaker.

**Comment by iopq** [comment]
@pliuphys according to leela zero chess threads, each network has its own "optimal" parameter set, I'm not sure if it applies to leela zero for go

**Comment by wonderingabout** [comment]
@croosn big thanks for the sgf

copy viewable here : https://online-go.com/game/16915236

**Comment by lightvector** [comment]
> That's a good point. In my fix I wasn't planning to ban broken ladders; just to better inform the tree search that they exist.
> 
> If we add input layers that point out when a chain is in inescapable atari, when a move puts a chain into inescapable atari, or when a move threatens to bring a chain out of inescapable atari, the network can still choose to run out with them.

@featurecat - You're also missing the case where a distant move makes a ladder work instead of making it not work (i.e. a ladder *maker* instead of a ladder *breaker*).

I recommend something a little more general than explicitly telling the network moves that change ladder working/not-working status - simply tell the network what 1-liberty stones where in inescapable atari or what 2-liberty stones were puttable in inescapable atari in one move, as of the current board position and as of the board position on the several previous turns. The network can then compute any diffs it likes, including understanding fights where the last several moves in a row have alternately broke and remade a distant ladder.

**Comment by featurecat** [comment]
I was thinking of modifying LZ at the tree search level. But I suppose I can also modify it at the network level and train on existing games, that might be worth trying. I'll think about it. It's hard to figure out which of these approaches is best. 

@gcp Have you been reading this thread? What are your thoughts?

**Comment by wonderingabout** [comment]
i'm not @gcp, but i wanted to say : i think its pointless to start a new run as long as the problematic blind spots in current engine code are not fixed, the new AI will still be defective

**Comment by Glrr** [comment]
Adding an input plane with a ladder feature would certainly be interesting, but a liberty feature would be as well. I already found a couple of blind spots of the network during games reviewing, and they are mostly related to liberties.

**Comment by gcp** [comment]
>The current "zero" bots are not really zero. They omit features like liberties, legality and ladder, but at the same time have the "age" feature for stones, encoded in a way that maximises the NNs involvement with it. The idea that a bot should focus attention near fresh stones so the output for the same board position can be different depending on this move order data - this is a human idea. 

The reason the network gets the previous positions is that it is required to see ko, which is in the game rules and so required even in a Zero approach. It has nothing to do with human ideas about "focusing attention".

**Comment by gcp** [comment]
>I don't know why this behavior happens. It's always obvious after a few thousand playouts the move with the higher win rate is better, but it keeps going with its original guess anyway
>Why doesn't it just search the move with the better win rate more?

The current search is the optimal one that has been found so far, so the answer to your question is generally "because it makes the program weaker".

**Comment by gcp** [comment]
>@gcp Have you been reading this thread? 

Mostly, no, see: https://github.com/leela-zero/leela-zero/issues/2157#issuecomment-456516287

>What are your thoughts?

1) I think for most people involved at this point a pure "zero" approach for a second run is not necessarily the most interesting nor the most useful.

2) I do think that if one adds features to the input, some consideration should be give to avoiding bias (the nice thing about zero is that it has none!). To understand what I mean here, consider giving liberty bitmaps (rather than floats or similar). If you add one for one liberty, that's essential enough that there's no bias. If you add 5 that go from 1,2,3,4,5 liberties, you've just introduced a discontinuity that might mislead how the network seeks the life and death of large groups.

Ladder features are IMHO tricky to do without this kind of bias. I don't think ladder reading is, in the general sense, a tractable problem. This means the ladder reading code and associated inputs will make mistakes, and if the network relies on them working, it might be very hard for it to overcome those.

I suspect that with SENets or liberty inputs, the ability to see "across the board" and detect ladders might be improved enough this is not a big issue any more. And I think such an approach is more free of bias than trying to write a perfect ladder detector. Maybe there's other bias-free tricks that would help ladder reading.

**Comment by kuba97531** [comment]
> The reason the network gets the previous positions is that it is required to see ko, which is in the game rules and so required even in a Zero approach (you can't understand what a legal move is without it). It has nothing to do with human ideas about "focusing attention".

I don't think it's true. First of all to know about the ko it's enough to know one board intersection that is a forbidden ko capture instead of last 8 moves. Actually I strongly believe that having the history is important aspect of playing strong games against equal opponnents.

It introduces a clear bias in LZ to answer moves locally. For example in this pair of positions: https://gyazo.com/14be0c2753f1c3c4d76a90112ba6cfd4

On the other hand being able to ask a program to evaluate position ignoring history would be a good option for game analysis.

> I think for most people involved at this point a pure "zero" approach for a second run is not necessarily the most interesting nor the most useful.

Absolutely. In my opinion the zero-approach has been proven to work and I would welcome the switch of the main goal of the project from getting the strongest zero-approach bot into making the engine most useful for humans.

**Comment by gcp** [comment]
>First of all to know about the ko it's enough to know one board intersection that is a forbidden ko capture instead of last 8 moves.

Tromp Taylor uses positional superko. In theory you need the entire board history, but of course that is prohibitive.

>It introduces a clear bias in LZ to answer moves locally. 

Yes, but this bias is learned by the program itself. It is a strange one, and suggests it is a shortcut taken by the NN to save "connections" for other features. But game history is part of the game rules, so this is a not a "human" "non-zero" bias.

>On the other hand being able to ask a program to evaluate position ignoring history would be a good option for game analysis.

You can wipe the history during training, or use the Leela Chess trick of copying the *current* position into the history.

**Comment by lightvector** [comment]
> > First of all to know about the ko it's enough to know one board intersection that is a forbidden ko capture instead of last 8 moves.
> 
> Tromp Taylor uses positional superko. In theory you need the entire board history, but of course that is prohibitive.
> 
> > It introduces a clear bias in LZ to answer moves locally.
> 
> Yes, but this bias is **learned by the program itself**. It is a strange one, and suggests it is a shortcut taken by the NN to save "connections" for other features. But game history is part of the game rules, so this is not a "human" "non-zero" bias. The program decided to make the optimization. Despite this, it still tenukis much stronger than humans, AFAIK.

Still, citing positional superko move legality as a reason to include more than 1 history plane is pretty funny given that cycles of length > 2 barely affect the strategy of the game at all compared to cycles of length 2, and where we know that by far the largest *effect* of the history planes has nothing to do with superko at all, but because the neural net will use the history planes to bias and improve its prediction in a massive number of non-ko-related positions. Although game theoretically it shouldn't matter, given limited capacity it *is* very useful information to be able to tell where the previous moves were when they were likely played by an agent much stronger than the net itself (namely, the net+search).

It's sort of funny because if we were to use an alternative ruleset or a different game with similar mechanics to Go but where game history was unnecessary for move legality or game result determination (or, game history beyond the immediate previous turn), we would likely want to still include history planes (or, history planes beyond the immediate previous turn) to allow the net to bias itself in exactly the same way to improve its strength. But we would have lost the excuse of move legality.

This is not at all an argument against anything @gcp has said and trying to claim that history is non-"zero", because history does clearly makes sense for Go. It's more an observation that in Go (and Chess) there's sort of a happy excuse to do something would have been more non-"zero" in some other games, which mainly shows how the spirit of "zero" in some ways can start to be slightly fuzzy and arbitrary at the boundaries. As we've also seen in a few other cases as well.

**Comment by gcp** [comment]
I agree with you. If the information wouldn't be useful, the network would not have figured out the trick.

I didn't realize that it may be beneficial for chess as well. Has someone measured that? I'd have expected a negligible effect, but you make me wonder...

**Comment by lightvector** [comment]
I haven't measured it of course, I'm just guessing, although perhaps someone from LC0 has done some experiments. It's probably far less important there than in Go, due to the smaller board size making vastly less important for the neural network to have some indication of what aspect of the board is now important to pay attention via seeing where previous moves were. But I would be very surprised it it didn't, say, produce a statistically significant improvement in policy predictive accuracy at least.

**Comment by killerducky** [comment]
See here I copied some old Lc0 results without history from Discord into this github isssue:
https://github.com/glinscott/leela-chess/issues/443#issuecomment-472523089

**Comment by alreadydone** [comment]
> I suspect that with SENets or liberty inputs, the ability to see "across the board" and detect ladders might be improved enough this is not a big issue any more. And I think such an approach is more free of bias than trying to write a perfect ladder detector. Maybe there's other bias-free tricks that would help ladder reading.

I am not seeing how SEnet would help with ladders, but I agree liberties would help. I think liberty features mainly serve as a low-liberty warning so only up to 3 liberties would be good enough (an isolated stone has only 4 liberties). Life-and-death of large groups may involve cut and connection (possibly by ko), and capturing races with eyes, which are too complicated to be captured by liberties alone anyway.

For reference KataGo uses 3 planes for 1,2,3 liberties, AG rollout policy uses 3 planes for 1,2,>=3, and AG policy/value nets use 8 planes for 1,2,...,7,>=8.
These features allow the network to see a long chain in atari, pursuing/escaping moves in a ladder, captures by throwing in an eye, snapbacks, etc.
As @gcp said, these features are free (already calculated to enable update of board state, no extra CPU burden for possibly branched ladder search), and I'm eager to see them used for self-play soon with net2net'd latest weights.

**Comment by lightvector** [comment]
I can affirm that liberty features act as a low-liberty warning, and primarily act to make the neural net never have blind spots as in https://github.com/leela-zero/leela-zero/issues/877. For example in that situation neural nets with a 1-liberty feature should put almost 100% mass on the black capture move, and having a 2-liberty feature will able to statically tell that white's move must be self-atari on a huge group, and therefore put almost no policy mass on that move. But they also don't magically solve ladder evaluation issues. The main difficulty is not perceiving that the ladder is in atari, it's in telling whether it eventually works or not.

Features for much larger numbers of liberties are much less useful, because groups are not always solidly connected, throwins and approach moves and ko mouths and and all sorts of other stuff affect the number of liberties for fighting purposes, etc. When I was doing supervised learning on pro games, I used to have binary features for 1,2,3,4 liberties, and I found that zeroing out the 4-liberty planes in the input *without re-training the neural net* only mildly harmed the policy loss (only a few centilogits), indicating that the neural net had already decided on its own not to pay much attention to them! So I eventually got rid of the 4-liberty planes, with basically no detectable harm on the pro move prediction accuracy when trained from the start without them.

**Comment by tapsika** [comment]
> > The current "zero" bots are not really zero. They omit features like liberties, legality and ladder, but at the same time have the "age" feature for stones, encoded in a way that maximises the NNs involvement with it. The idea that a bot should focus attention near fresh stones so the output for the same board position can be different depending on this move order data - this is a human idea.
> 
> The reason the network gets the previous positions is that it is required to see ko, which is in the game rules and so required even in a Zero approach (you can't understand what a legal move is without it). It has nothing to do with human ideas about "focusing attention".

This is why I wrote: 

> > (Even if you think this is necessary for detecting long cycles, in practice it is mostly used for other things.)

As for "nothing to do" with it, I also think the practical reality is:

> > It introduces a clear bias in LZ to answer moves locally.
> 
> Yes, but this bias is **learned by the program itself**. It is a strange one, and suggests it is a shortcut taken by the NN to save "connections" for other features. But game history is part of the game rules, so this is not a "human" "non-zero" bias.

The same logic would go for a liberty input as well (capture rule). And game history is zero, but move legality is non-zero? This seems like a contradiction.

In any case there are significant grey zones about the zero idea, and as others said it may not be that important or interesting now for the future of open source go. Also the main issue for well trained nets is ladder (guessing on sight is esp. important at deeper levels in the search tree), and it is clear that even an intermediate solution like "color and distance of closest stone in all directions" would not be zero.

**Comment by gcp** [comment]
>See here I copied some old Lc0 results without history from Discord into this github isssue:
glinscott/leela-chess#443 (comment)

100 Elo? Holy smokes!

**Comment by gcp** [comment]
>I am not seeing how SEnet would help with ladders

You would get wider propagation that a large group is in trouble. But indeed that info is too generic to be useful near the ladder-stopper again :-/

**Comment by alreadydone** [comment]
> 100 Elo? Holy smokes!

So removing randomness in move selection has the good side effect of producing higher quality history ...
